# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [156]:
import warnings
warnings.simplefilter('ignore')
import sqlite3
import pandas as pd
import numpy as np

Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [157]:
database_connection = sqlite3.connect('go_sales.sqlite')

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [158]:
product = pd.read_sql_query("SELECT * FROM product", database_connection)
product_type = pd.read_sql_query("SELECT * FROM product_type", database_connection)
product_line = pd.read_sql_query("SELECT * FROM product_line", database_connection)
sales_staff = pd.read_sql_query("SELECT * FROM sales_staff", database_connection)
sales_branch = pd.read_sql_query("SELECT * FROM sales_branch", database_connection)
retailer_site = pd.read_sql_query("SELECT * FROM retailer_site", database_connection)
country = pd.read_sql_query("SELECT * FROM country", database_connection)
order_header = pd.read_sql_query("SELECT * FROM order_header", database_connection)
order_details = pd.read_sql_query("SELECT * FROM order_details", database_connection)
sales_targetdata = pd.read_sql_query("SELECT * FROM sales_targetdata", database_connection)
returned_item = pd.read_sql_query("SELECT * FROM returned_item", database_connection)
return_reason = pd.read_sql_query("SELECT * FROM return_reason", database_connection)

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [94]:
product['PRODUCTION_COST'] = pd.to_numeric(product['PRODUCTION_COST'])
product_kosten_tussen_100_en_50 = (product['PRODUCTION_COST'] < 100) & (product['PRODUCTION_COST'] > 50)
product.loc[product_kosten_tussen_100_en_50, ['PRODUCT_NUMBER', 'PRODUCTION_COST']]     

PRODUCT_NUMBER  PRODUCTION_COST
13            110            60.00
14            111            81.80
17            114            80.00
24             17            60.00
25             18            86.00
28             20            78.55
31             23            65.33
32             24            62.50
56             46            52.86
58             48            55.00
69             58            57.13
71              6            85.11
81             69            73.33
83             70            80.00
84             71            60.00
85             72            72.50
88             75            80.00
91             78            79.19
92             79            92.58
93              8            75.00
94             80            94.12
97             83            56.25
98             84            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [95]:
product['MARGIN'] = pd.to_numeric(product['MARGIN'])
product_margin_buiten_20_en_60 = (product['MARGIN'] < 0.20) | (product['MARGIN'] > 0.60)
product.loc[product_margin_buiten_20_en_60, ['PRODUCT_NUMBER', 'MARGIN']]    

PRODUCT_NUMBER  MARGIN
29              21    0.17
32              24    0.17
100             86    0.67
101             87    0.67
102             88    0.70
103             89    0.63
105             90    0.63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [96]:
country_currency_name_francs = country['CURRENCY_NAME'] == 'francs'
country.loc[country_currency_name_francs, ['COUNTRY']].sort_values(['COUNTRY'])

COUNTRY
7       Belgium
0        France
19  Switzerland

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [97]:
product['MARGIN'] = pd.to_numeric(product['MARGIN'])
product_introductiedatums_margin_meer_50 = (product['MARGIN'] > 0.50)
product.loc[product_introductiedatums_margin_meer_50, ['INTRODUCTION_DATE']].drop_duplicates(['INTRODUCTION_DATE'])

INTRODUCTION_DATE
4          10-12-2019
6          18-12-2019
7          27-12-2019
9           13-1-2020
14         15-12-2019
79         26-10-2016
100         15-2-2011

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [38]:
sales_branch_adres_regio = (sales_branch['ADDRESS2'].notna()) & (sales_branch['REGION'].notna())
sales_branch.loc[sales_branch_adres_regio, ['ADDRESS1', 'CITY']]

ADDRESS1              CITY
10  Prol. Paseo de la Reforma No. 51  Distrito Federal
11               202-2-3 Hyakunincho             Tokyo
12                     543-225 Asahi        Osaka City
13                  2315 Queen's Ave         Melbourne
15             Avenida Paulista, 333         São Paulo
21                    3 Albany Court        Birmingham
27               Singelgravenplein 4         Amsterdam

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [43]:
country_dollars = (country['CURRENCY_NAME'] == 'dollars') | (country['CURRENCY_NAME'] == 'new dollar')
country.loc[country_dollars, ['COUNTRY']].sort_values(['COUNTRY'])

COUNTRY
6       Australia
15         Canada
3          Taiwan
14  United States

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [50]:
retailer_site_adressen_stad_postcode_D = (retailer_site['POSTAL_ZONE'].str[0] == 'D') & (retailer_site['ADDRESS2'].notna())
retailer_site.loc[retailer_site_adressen_stad_postcode_D, ['ADDRESS1', 'ADDRESS2', 'CITY']].sort_values(['ADDRESS2'])

ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [89]:
returned_item["RETURN_QUANTITY"] = returned_item["RETURN_QUANTITY"].astype(int)
returned_item["RETURN_QUANTITY"].sum()

14306

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [102]:
retailer_site['REGION'].count()

251

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [184]:
product['MARGIN'] = pd.to_numeric(product['MARGIN'], errors='coerce')

margin_low = product['MARGIN'].min()
margin_high = product['MARGIN'].max()
margin_avg = product['MARGIN'].mean().round(2)
print_margins = margin_low, margin_high, margin_avg
print_margins

(0.17, 0.7, 0.4)

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [133]:
retailer_site['CITY'] = (retailer_site['ADDRESS2'] == 'NULL')
retailer_site['CITY'].count()

391

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [155]:
order_details['UNIT_COST'] = (order_details['UNIT_SALE_PRICE']) < (order_details['UNIT_PRICE'])
order_details['UNIT_COST'].mean()

0.32575528876297516

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [182]:
aantal_per_functie = sales_staff.groupby('POSITION_EN')['SALES_STAFF_CODE'].count()
aantal_per_functie.columns = ['SALES_STAFF_CODE', 'POSITION_EN']
aantal_per_functie

POSITION_EN
Branch Manager                  19
District Manager                 4
General Manager                  1
Level 1 Sales Representative    13
Level 2 Sales Representative    39
Level 3 Sales Representative    24
Regional Manager                 2
Name: SALES_STAFF_CODE, dtype: int64

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [190]:
aantal_per_telefoonnummer = sales_staff.groupby('WORK_PHONE')['SALES_STAFF_CODE'].count()
aantal_per_telefoonnummer = aantal_per_telefoonnummer[aantal_per_telefoonnummer > 4]
aantal_per_telefoonnummer.columns = ['WORK_PHONE', 'SALES_STAFF_CODE']
aantal_per_telefoonnummer

WORK_PHONE
+(44) 121 3505267      5
+(61) 03 2982  4242    5
+31 (0)20 692 93 94    5
+33 1 68 94 52 20      5
1 (206) 292-0012       5
1 (305) 557-4810       6
1 (310) 281-5722       5
1 (403) 232-5986       5
1 (416) 493-5595       5
1 (617) 268-6754       6
Name: SALES_STAFF_CODE, dtype: int64

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [90]:
merged_product_and_product_type.groupby('PRODUCT_TYPE_EN', as_index=False)['PRODUCT_NUMBER'].count()

NameError: name 'merged_product_and_product_type' is not defined

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [1]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [2]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 